torch/csrc/jit/interpreter.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+#include <memory>
+#include <vector>
+
+namespace at {
+  struct Tensor;
+}
+namespace torch { namespace jit {
+
+struct NotImplementedException : public std::logic_error {
+  NotImplementedException()
+  : std::logic_error("Function not yet implemented.") {}
+};
+
+// The interpreter run Graphs with Tensor inputs and Tensor outputs
+// a separate component in the autograd handles unwrapping and wrapping
+// variable objects for use in the interpreter.
+
+struct FunctionImpl;
+struct InterpreterImpl;
+struct Graph;
+
+struct Function {
+  Function()
+  : pImpl(nullptr) {}
+  Function(std::shared_ptr<Graph> & graph);
+  ~Function();
+  operator bool() const {
+    return pImpl != nullptr;
+  }
+private:
+  std::shared_ptr<FunctionImpl> pImpl;
+  friend class InterpreterImpl;
+};
+
+struct Interpreter {
+  Interpreter(const Function & function);
+  void runOneStage(
```

Comment:
You can tell this is the case when you look at the 'clone' method below, but there should be comments here saying that an interpreter is a mutable resource which is not thread safe and can only be run once. Perhaps there is another name rather than Interpreter which expresses this more clearly, like "InterpreterHandle"

Reply:
I'll add some comments. I will probably rename Interpreter -> InterpreterState to make it more clear. `lua_State` is a similar object.